indico/util/locales.py

```diff
@@ -0,0 +1,63 @@
+# This file is part of Indico.
+# Copyright (C) 2002 - 2024 CERN
+#
+# Indico is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see the
+# LICENSE file for more details.
+
+from functools import cached_property
+
+from indico.core.config import config
+
+from .i18n import get_all_locales
+
+
+class LocaleChoice:
+    def __init__(self, forced_context_locale, additional_context_locales, user_locale,
+                 user_forces_own_locale):
+        self.forced_context_locale = forced_context_locale
+        self.additional_context_locales = additional_context_locales or []
+        self.user_locale = user_locale or config.DEFAULT_LOCALE
+        self.user_forces_own_locale = user_forces_own_locale
+
+    @cached_property
+    def current_locale(self):
+        return self.forced_context_locale or self.user_locale
+
+    @cached_property
+    def context_forces_locale(self):
+        return self.forced_context_locale is not None
+
+    @cached_property
+    def has_restircted_choice(self):
+        return not self.user_forces_own_locale and self.context_forces_locale
+
+    @cached_property
+    def has_no_choice(self):
+        return self.has_restircted_choice and len(self.additional_context_locales) < 2
+
+    @cached_property
+    def all_locales(self):
+        return get_all_locales()
+
+    @cached_property
+    def all_locales_sorted(self):
+        return sorted(self.all_locales)
+
+    @cached_property
+    def selectable_locales(self):
+        if self.has_no_choice:
+            return [self.forced_context_locale]
+        if self.has_restircted_choice:
+            return self.additional_context_locales
+        return self.all_locales
+
+    def get_locale_info(self, locale):
+        fallback_locale_info = self.all_locales.get(config.DEFAULT_LOCALE)
+        return self.all_locales.get(locale, fallback_locale_info)
+
+    def get_locale_display(self, locale):
+        (name, territory, is_ambiguous) = self.get_locale_info(locale)
+        if is_ambiguous:
+            return f'{name} ({territory})'
```

Comment:
That actually means "Portugese" *is* ambiguous, no?

Reply:
That is, it's a data issue, not UI issue.

Reply:
So you currently use `is_ambiguous` to mean "We have more than 1 translation for languages with the same name", but it actually should mean "We have more than 1 language with the same name (regardless of whether there are translation for it)". It's a domain issue, not a presentation issue.

Reply:
ah, i didn't realize `is_ambiguous` isn't new but something we already had. Now that I noticed it and checked how we're using it I remember why it's there:

When showing the current language I wanted to avoid being overly verbose except when it's needed, but when showing the languages to choose from I'm always showing the full name with the territory included. That way you know what you get when changing the language, but once you are using the language we prefer to show just the short name - unless Indico has more than one version of that language.

Reply:
Ok

Reply:
In that case, I think you don't really need the `is_ambiguous` flag at all?